### ISSUE TITLE
New must_not_suspend lint triggers even when value has been moved

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust#89562]
I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d4a6d29686fc96f1a1846e3fe35b2109)):

```rust
use std::sync::Mutex;

async fn foo() {
    let foo = Mutex::new(1);
    let lock = foo.lock().unwrap();

    // Prevent mutex lock being held across `.await` point.
    drop(lock);

    bar().await;
}

async fn bar() {}
```

I expected to see this happen: No warnings (except unused fns)

Instead, this happened: `must_not_suspend` triggered

### Meta

`rustc --version --verbose`:
```
rustc 1.57.0-nightly (003d8d3f5 2021-10-04)
binary: rustc
commit-hash: 003d8d3f56848b6f3833340e859b089a09aea36a
commit-date: 2021-10-04
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

**Comment by jplatte** [comment]
When I encountered this in the real world, the `drop` and subsequent `.await` was in a branch so using a scope instead would have involved extra control flow. I silenced the lint for now.

**Comment by memoryruins** [comment]
The lint seems to be correct currently, since using `drop` is not enough to make the future `Send`.

```rust
fn is_send<T: Send>(_x: T) {}

// error: future cannot be sent between threads safely
fn main() { is_send(foo()); }
```

Wrapping in a block will avoid this issue for now like in https://github.com/rust-lang/rust/issues/57478#issuecomment-501186084 until there are more precise generator captures https://github.com/rust-lang/rust/issues/69663

**Comment by jplatte** [comment]
In my case whether the future is `Send` or not didn't seem to matter (for whatever reason), avoiding possible deadlocks seemed good enough. Maybe using an async-aware mutex would be the best solution though 🤔

**Comment by jplatte** [comment]
Wait it says that *I* closed this issue. If I did, that wasn't intentional so I'll reopen...
